### ISSUE TITLE
Different accuracy for distributed horovod vs. single machine

### ISSUE THREAD

**Issue by aaron276h** [uber/horovod#390]
I have observed that distributed horovod  reaches lower accuracy than single machine horovod when using the same hyperparameter (e.g., same global batch size, same learning rate, same amount of training).

The set-up I am using is a two machines with one gpu each for the distributed version, and one machine with one gpu for the single machine version. I am running this using the Horovod docker containers. 

After 5 epoch of training ResNet-50 w/ global mini-batch size of 64, and consistent learning rate of 0.01, the single machine case converges to validation Accuracy @ 1 = 0.4912 Accuracy @ 5 = 0.7495, while the distributed case converges to Accuracy @ 1 = 0.4541 Accuracy @ 5 = 0.7166. I have verified that these numbers are consistent across multiple runs. I have found that distributed TF (using PS) has the same issue but not as drastically, with the distributed version achieving Accuracy @ 1 = 0.4655 Accuracy @ 5 = 0.7276 after 5 epochs.

To check if this is related how momentum is applied, I also ran the experiments using SGD, and found the discrepancy to be even larger, with single machine achieving Accuracy @ 1 = 0.3827 Accuracy @ 5 = 0.6442 and the distributed case Accuracy @ 1 = 0.3196 Accuracy @ 5 = 0.5813. 

For completness I am attaching the code I used to trigger the runs.

Single machine:
`mpirun -np 1 -mca plm_rsh_args "-p 12345" -x LD_LIBRARY_PATH -x PATH  -H host0:1 python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=64 --model=resnet50 --variable_update=horovod --data_name=imagenet --data_dir=$data/imagenet --init_learning_rate=0.01000 --optimizer=momentum --num_epochs=5 --train_dir=$train_dir/single-horovod --summary_verbosity=1 --horovod_device cpu`

Distributed:
`mpirun -np 2 -mca plm_rsh_args "-p 12345" -x LD_LIBRARY_PATH -x PATH  -H host0:1,host1:1 python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=32 --model=resnet50 --variable_update=horovod --data_name=imagenet --data_dir=$data/imagenet --init_learning_rate=0.01000 --optimizer=momentum --num_epochs=5 --train_dir=$train_dir/distributed-horovod --summary_verbosity=1 --horovod_device cpu`

Testing:
`mpirun -np 1 -mca plm_rsh_args "-p 12345" -x LD_LIBRARY_PATH -x PATH  -H host0:1 python tf_cnn_benchmarks.py --eval=True --num_gpus=1 --batch_size=100 --model=resnet50 --variable_update=horovod --data_name=imagenet --data_dir=$data/imagenet --num_epochs=1 --train_dir=$train_dir/$experiment --summary_verbosity=1 --horovod_device cpu`

**Comment by tgaddair** [comment]
Alex may have some additional thoughts, but those results seem about in line with what I'd expect given how the gradient averaging works.

![gradient_averaging](https://user-images.githubusercontent.com/1742912/42957524-cfb23078-8b37-11e8-9336-0ba7aa11d161.png)

In the single-threaded world, you process once batch at a time, compute the model updates (gradients), and apply them to update the weights.  But with distributed training, you process N batches at once, then average the gradients together, which is not quite the same process.  

So I wouldn't be surprised if it takes more epochs to reach the same convergence, but the theory is that the decrease in per-batch loss improvement should be significantly less than the efficiency gains of running many batches in parallel.

**Comment by aaron276h** [comment]
@tgaddair Thanks for the quick reply! So if there is averaging of gradients going on this result does appear to make sense. My follow up questions is why doesn't Horovod just sum the gradients instead? This would appear to allow it to have the quality-of-work as a single threaded execution. Also would you know how difficult it would be to make this change? 

P.S To double check that we are talking about check that we are talking about gradient averaging the same way. The way horovod does averaging is by summing up the gradients from each training process and divide by the number of training processes.

**Comment by alsrgv** [comment]
@aaron276h, I've noticed you're using batch size of 16 for distributed case and batch size of 64 for the single GPU case. Can you clarify the reasoning behind it?

The benchmarking code actually does simply sum up gradients.

**Comment by tgaddair** [comment]
I believe that whether you sum or average the gradients, there will necessarily be some differences.  In the parallel case, batches 1 and 2 will compute their gradients using the same initial set of parameters, but in the serial case, batch 2 will be using updated parameters obtained as a result of batch 1's updates.  So the gradients are expected to differ.

Though it sounds like averaging isn't occurring in this particular case, the rationale behind doing it can be found in a paper published by Facebook: [Accurate, Large Minibatch SGD:
Training ImageNet in 1 Hour](https://arxiv.org/pdf/1706.02677.pdf).

Hope that helps!

**Comment by aaron276h** [comment]
@alsrgv Sorry that was typo, just double checked my experiment logs and I did in fact use 64 in the single machine case, and 32 per machine in the 2 machine distributed case. (Just updated my original post to correct the error). 

@tgaddair I agree with you that if you keep the mini-batch per machine size consistent then the summing would lead to a different result. However if you keep the global mini-batch size consistent, I believe single worker case should be the same as distributed with gradients being summed.

**Comment by alsrgv** [comment]
@aaron276h, got it, thanks for clarifications. Part of the difference could be because of different batch size and effect of that on batch normalization. In order to get speedup in convergence, you'd actually want to keep batch size per GPU the same - 64.  This increases total effective batch size, but adding up gradients compensates for that. Can you try that? Also, five epoch mark may be not the best place to compare final accuracy. Usually the final convergence is achieved after 90 epochs for resnet models.

**Comment by aaron276h** [comment]
@alsrgv Yes I agree batch normalization could definitely be the cause of this difference. Although I am unsure then of why TF+PS performs differently than Horovod. Would you have any insight into this?

I also agree that five epochs might be too early to compare final accuracy, it just happens to be suitable for the use-case we are trying to pursue, which is why it was the testing point. 

I will run with the distributed case with per GPU size of 64 (scaling the LR by 2x) and post the results.

**Comment by alsrgv** [comment]
@aaron276h, since the benchmarks are summing up the gradients, I don't think you should adjust the learning rate parameter.

**Comment by aaron276h** [comment]
@alsrgv I ran both with and without learning rate adjustment, although if the benchmark is indeed just summing them and never diving by number of batches, learning rate adjustment should not be needed.

2 Machine, global batch size 128, lr = 0.01 after 5 epochs: Accuracy @ 1 = 0.4331 Accuracy @ 5 = 0.6984

2 Machine global batch size 128, lr = 0.02 after 5 epochs: Accuracy @ 1 = 0.4543 Accuracy @ 5 = 0.7184

**Comment by alsrgv** [comment]
@aaron276h, I really think you need to train till convergence to get a feel for final results.

For your reference, here are three runs of PyTorch ImageNet example with identical parameters:
![image](https://user-images.githubusercontent.com/16640218/43032188-43f7d4ae-8c66-11e8-8b80-1e8aab64ffbf.png)

On the low epoch count, we see variability of 40-46%, but they converge to the same final result.

**Comment by aaron276h** [comment]
@alsrgv Yep that definitely makes sense, we are going to run our different branches to completion and see if we get the same results. Just curious, the results you posted, is that executed on different number of machines or same configuration every time?

**Comment by alsrgv** [comment]
@aaron276h, the same configuration every time.

**Comment by zenetio** [comment]
I executed the same command line with 1 and 4 GPUs and the accuracy with 1 GPU is better than with 4 GPUs, although the training time with 4 is almost 50% faster. Maybe you need decide if the lower in accuracy may justify the decrease in training time.
